torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2003,6 +2014,300 @@ def clip_grad_norm_(
                 assert p.grad is not None
                 p.grad.detach().mul_(clip_coef.to(p.grad.device))
 
+    @staticmethod
+    def full_optim_state_dict(
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        optim_input: Optional[Union[
+            List[Dict[str, Any]], Iterable[torch.nn.Parameter],
+        ]] = None,
+    ) -> Dict[str, Any]:
+        """
+        Consolidates the full optimizer state on rank 0 and returns it
```

Comment:
nit: please ignore if this ask does not make sense. If this makes sense, it's totally fine to do it in followup PRs. 

This looks like a gather op. For the gather in c10d, there is an optional `dst=0` argument. Does it make sense to have it here as well?

Reply:
This currently uses `_all_gather_base()` because I was unable to get `gather()` to work correctly. I can look into trying `gather()` again in a follow up.

Regarding specifying a target rank, I no longer included that argument and simply have a global `OPTIM_TARGET_RANK = 0` at the top of `optim_utils.py`, which I meant to unify with model state dict in the future. I think we simply need some design consensus on this.
cc: @rohan-varma @zhaojuanmao
In my understanding, currently full model state dict saves on all ranks since it does not pass `rank0_only=True` to `summon_full_params()`. Will there be a change to only save it on rank 0 by default?

Either way, there will certainly be a follow-up to unify with model state dict since the plan is to pass a config into `state_dict_type()` to choose settings like whether to offload to CPU or not. Those settings will equally configure the optimizer state checkpointing.

Reply:
Looks like others in OSS are experiencing correctness issues with `dist.gather()` as well: https://github.com/pytorch/pytorch/issues/74809